core/near-network/src/lib.rs

```diff
@@ -0,0 +1,400 @@
+use ::primitives::types::AccountId;
+use ::primitives::hash::CryptoHash;
+use ::primitives::traits::{Encode, Decode};
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::net::SocketAddr;
+use std::io::{Error, ErrorKind};
+use ::tokio::net::{TcpStream, TcpListener, tcp::ConnectFuture};
+use ::tokio_codec::{Framed};
+use ::tokio::prelude::stream::SplitStream;
+use ::futures::sync::mpsc::{channel, Sender};
+use ::futures::{Poll, Stream, Async, Future, Sink};
+use ::bytes::Bytes;
+use ::log::{debug, error};
+use ::serde_derive::{Serialize, Deserialize};
+use ::rand::{thread_rng, seq::IteratorRandom};
+
+mod codec;
+use crate::codec::Codec;
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug, Serialize, Deserialize)]
+pub struct Peer {
+    /// address of peer
+    addr: SocketAddr,
+    /// peer id
+    id: PeerId,
+    /// account id, peer may not always have one
+    account_id: Option<AccountId>,
+}
+
+impl Peer {
+    pub fn new(addr: SocketAddr, id: PeerId, account_id: Option<AccountId>) -> Self {
+        Peer {
+            addr,
+            id,
+            account_id
+        }
+    }
+}
+
+/// Handshake info we exchange when first establish connection with a peer. 
+// TODO: make it secure
+#[derive(Serialize, Deserialize)]
+pub struct HandShake {
+    peer_id: PeerId,
+    account_id: Option<AccountId>,
+}
+
+/// unique identifier for nodes on the network
+// Use hash for now
+pub type PeerId = CryptoHash;
+
+// This is very awkward because we cannot copy sink and have to use a channel
+// to tunnel the data through. Definitely need to revisit this later to check for
+// better solutions
+type NearStream = (Sender<Bytes>, SplitStream<Framed<TcpStream, Codec>>);
+
+/// convert raw TcpStream to NearStream
+fn convert_stream(stream: TcpStream) -> NearStream {
+    let (sender, receiver) = channel(1024);
+    let (sink, stream) = Framed::new(stream, Codec::new()).split();
+    // spawn the task that forwards what receiver receives to send through sink
+    tokio::spawn(
+        sink.send_all(
+            receiver.map_err(|e| Error::new(ErrorKind::BrokenPipe, format!("{:?}", e)))
+        )
+        .map(|_| ())
+        .map_err(|e| error!("Error when forwarding: {:?}", e))
+    );
+    (sender, stream)
+}
+
+pub struct Service {
+    /// peer id of the node
+    peer_id: PeerId,
+    /// account id of the node
+    account_id: Option<AccountId>,
+    // TODO: listen on multiple address
+    listener: TcpListener,
+    /// Local info about accounts
+    account_to_peer: HashMap<AccountId, Peer>,
+    /// peers that we are trying to connect to 
+    pending_peers: HashSet<SocketAddr>,
+    /// pending connection futures
+    pending_connections: VecDeque<(SocketAddr, ConnectFuture)>,
```

Comment:
As we discussed, having our own pool of pending futures that we explicitly iterate with some task is an exact re-implementation of the core tokio functionality that works with kqueue/epoll. We should not be reimplementing it, because it will be highly inefficient and will require more code than just using tokio.

If we are trying to send a message to a peer then:
* if we have open connection to this peer we just send it over. We spawn a task that puts message in the open socket;
* if we know ip:port of that peer we open said connection and send the message over. We spawn a task to do that;
* if we don't know ip:port we fail without panic.

If we receive an incoming socket connection we create:
* one task that listens to the incoming stream of messages and calls Protocol on it;
* another task that opens a channel and redirects messages from that channel into the sink of the socket;

We need only two more tasks:
* Task that periodically announces its known peer info to random peers;
* Task that listens to such announcements and merges the peer info.

Reply:
That is what I tried previously, but due to some unknown problem with lock/asynchronous code, it didn't work. It could be related to the poll method I implemented, will investigate more.

Reply:
You could submit it to a separate branch and we could debug it together.

Reply:
Fixed in 9896565